beluga/include/beluga/sensor/bearing_sensor_model.hpp

```diff
@@ -0,0 +1,208 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_SENSOR_LANDMARK_BEARING_MODEL_HPP
+#define BELUGA_SENSOR_LANDMARK_BEARING_MODEL_HPP
+
+// external
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/transform.hpp>
+#include <sophus/se2.hpp>
+#include <sophus/se3.hpp>
+
+// standard library
+#include <cmath>
+#include <complex>
+#include <random>
+#include <tuple>
+#include <vector>
+
+/**
+ * \file
+ * \brief Implementation of a discrete landmark bearing sensor model.
+ */
+
+namespace beluga {
+
+/// Parameters used to construct a BearingSensorModel instance.
+struct BearingModelParam {
+  double sigma_bearing{0.2};                           ///< Standard deviation of the bearing error.
+  Sophus::SE3d sensor_in_robot_footprint_transform{};  ///< Pose of the sensor in the robot reference frame.
```

Comment:
Agreed, `sensor_in_robot_footprint_transform` is too much.

Also :100: behind having transform naming convention, even if informal. Consistent naming goes a long way to make transform frame errors obvious when chaining transforms, inverting them, and making sure you are applying the right one (that's how I found the remaining one in the bearing model).

I like 2. In parts of the code I've shortened a similar nomenclature by only using the first letter of the parent frame (`robot_in_w_transform`, instead of `robot_in_world_transform`) since that's usually enough, but I don't feel too strongly about it.

Regarding 3, I can live with it, but I feel it more ambiguous than "in", though this may be because of my own mental framework to think about transforms: moving something "to" the parent frame is what the transform does to something else (a vector, another transform), but the transform itself represents the pose of the child frame "in" the parent frame.

The thing I don't like about 4 is that it's not obviously clear which is the parent frame and which the child. I've seen both options (parent-before and parent-behind) for the $_AT^B$ notation, so the interpretation will depend on the readers preferred literature.

In all cases I like the apparent simplicity of "frame" over "transform". The reason I used transform in many cases is because "robot_in_world_frame" could as well be assumed a position or a bearing, instead of a pose. So, I had "sensor_in_world_transform" and "landmark_in_world_bearing". I was not consistent about this, however.

Reply:
I updated this PR with the `<child>_in_<parent>_<type>` naming style. Check if you think that's something we're comfortable with.

Reply:
I'm like your proposal, but I'm not 100% in yet. How do you feel about `<child>_<type>_in_<parent>_frame`?

- sensor_pose_in_robot_frame
- robot_pose_in_world_frame
- detection_position_in_world_frame
- detection_position_in_sensor_frame
- detection_bearing_in_robot_frame
- world_pose_in_sensor_frame
- landmark_a_bearing_in_sensor_frame

Reply:
To add to the discussion, once you get used to [monogram notation](https://drake.mit.edu/doxygen_cxx/group__multibody__notation__basics.html), it is rather concise and unambiguous. I'm cool with whatever you pick though.

Reply:
> To add to the discussion, once you get used to [monogram notation](https://drake.mit.edu/doxygen_cxx/group__multibody__notation__basics.html), it is rather concise and unambiguous. I'm cool with whatever you pick though.

Not having gone too deep into it, I tend to think that's a bit too much. It looks powerful and probably makes sense for something large with complex physical quantities, but the main issue I see for our it requires external information to understand and use it. Something like `Qi_RT_F` and `w_AB` does not really mean anything to the non-initiated. Also, it clashes with the coding style for other variables.

> I like your proposal, but I'm not 100% in yet. How do you feel about <child>_<type>_in_<parent>_frame?

I'm ok with it, but I feel the last `_frame` can be elided. Like:

- sensor_pose_in_robot
- robot_pose_in_world
- detection_position_in_world
- detection_position_in_sensor
- detection_bearing_in_robot
- world_pose_in_sensor
- landmark_a_bearing_in_sensor

Reply:
> I'm ok with it, but I feel the last `_frame` can be elided.

Sounds good!

Reply:
done.